### PR TITLE
[FLINK-11156] [tests] Reconcile ZooKeeperCompletedCheckpointStoreMockitoTest with JDK 9

### PR DESCRIPTION
**Pull request by TisonKun**
## What is the purpose of the change

The issue occurs because powermock hasn't fully supported JDK 9+ yet. Specifically, when we `@PrepareForTest` on a class(`ZooKeeperCompletedCheckpointStore`) has a static final field initialized by `Comparator#comparing`, it causes `IncompatibleClassChangeError`. We can fallback to use the previous syntax as a workaround.

Given https://github.com/powermock/powermock/issues/901 the author of powermock said "Right now the project is in frozen state because I do not have enough time.", we'd better fix the issue different from waiting to bump version.

Here we have to use powermock for hijacking the constructor of `ZooKeeperStateHandleStore`. An alternative for resolving this issue is get rid of powermock. I have tried it but ended with adding a testing constructor of `ZooKeeperStateHandleStore`, which looks quite hack for me.

It may be better to introduce builders for classes depending on `powemock#whennew` so we can get rid of it.

## Brief change log

- fallback the implementation to previous style as a workaround

## Verifying this change

This pull request itself is the test. Run `mvn test -Pjava9 -DfailIfNoTests=false -Dcheckstyle.skip=true -Dtest=org.apache.flink.runtime.checkpoint.ZooKeeperCompletedCheckpointStoreMockitoTest` for testing locally.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

cc @GJL

### CODE REVIEW

**Comment by GJL** [comment]
Thank you for your contribution to Apache Flink @TisonKun 
I think an additional constructor annotated with `@VisibleForTesting` is more maintainable. The 
current solution has the following downsides:
- It is not obvious why `Comparator.comparing` does not work. The next person might try to refactor it back to the old solution and break java 9 compatibility.
- The test uses powermock which might break again in later java versions.

Let me loop in @tillrohrmann, the original author of the test. What do you think?

**Comment by TisonKun** [comment]
@GJL I can do it with a testing constructor. If it is the style that Flink would be happy with, I will send a commit for it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java [hunk 1]
```diff
@@ -327,7 +327,6 @@ public int exists(String pathInZooKeeper) throws Exception {
 	 * @return True if the state handle could be released
 	 * @throws Exception If the ZooKeeper operation or discarding the state handle fails
 	 */
-	@Nullable
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java]: unrelated, move into a separate hotfix commit

**Comment by TisonKun** [comment]
I've updated the pull request, and the final usage of mockito is to mock CuratorFramework. I'd like to listen to @tillrohrmann to learn if we can test with Curator without mockito. I am rethinking ZooKeeper based HaService and learn how to test would be quite helpful.

Thanks for your review @GJL !

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java [hunk 2]
```diff
@@ -128,7 +145,9 @@ public ZooKeeperCompletedCheckpointStore(
 		// All operations will have the path as root
 		this.client = client.usingNamespace(client.getNamespace() + checkpointsPath);
 
-		this.checkpointsInZooKeeper = new ZooKeeperStateHandleStore<>(this.client, stateStorage);
+		this.checkpointsInZooKeeper = (checkpointsInZooKeeper != null) ?
```
Comment (TisonKun) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java]: I am not super happy with this line. This is why I have given up this way the first time. I'd like to learn whether this style is OK in Flink or we tend to build such a testing constructor in another way.

File: flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreMockitoTest.java [hunk 3]
```diff
@@ -89,46 +77,72 @@ public void testCheckpointRecovery() throws Exception {
 		expectedCheckpointIds.add(1L);
 		expectedCheckpointIds.add(2L);
 
-		final RetrievableStateHandle<CompletedCheckpoint> failingRetrievableStateHandle = mock(RetrievableStateHandle.class);
-		when(failingRetrievableStateHandle.retrieveState()).thenThrow(new IOException("Test exception"));
-
-		final RetrievableStateHandle<CompletedCheckpoint> retrievableStateHandle1 = mock(RetrievableStateHandle.class);
-		when(retrievableStateHandle1.retrieveState()).then(
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStoreMockitoTest.java]: are all these refactorings necessary for the test to run on jdk 9? If not, please move them into a separate commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java]: it's a bit icky but non-intrusive enough to be acceptable imo,

**Comment by TisonKun** [comment]
I did it with minimum changes. If we prefer testing classes to mockito style, we can make a follow up commits, maybe a separated pull request. It's up to the community convention.

**Comment by TisonKun** [comment]
@GJL for "It is not obvious why Comparator.comparing does not work.", I did a survey and

it points to a call to static method of interface(`Comparator#comparing` here) while using previous version of ASM(of course mock framework hacking bytecodes).

Groovy community has suffered this issue. FYI, https://issues.apache.org/jira/browse/GROOVY-8338 https://issues.apache.org/jira/browse/GROOVY-8528

**Comment by GJL** [comment]
@zentol Do you want to review the PR again?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java [hunk 4]
```diff
@@ -107,11 +108,27 @@
...
+		CuratorFramework client,
+		String checkpointsPath,
+		RetrievableStateStorageHelper<CompletedCheckpoint> stateStorage,
+		Executor executor) throws Exception {
+		this(maxNumberOfCheckpointsToRetain,
+			client,
+			checkpointsPath,
+			stateStorage,
+			executor,
+			null);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java]: we could pass `new ZooKeeperStateHandleStore<>(this.client, stateStorage)` instead; then we wouldn't need the null check in the other constructor

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java [hunk 5]
```diff
@@ -123,12 +140,14 @@ public ZooKeeperCompletedCheckpointStore(
 
 		// Ensure that the checkpoints path exists
 		client.newNamespaceAwareEnsurePath(checkpointsPath)
-				.ensure(client.getZookeeperClient());
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java]: revert

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java (hunk 4 above)
Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java]: we don't initalize `this.client` in this constructor. and before construct `checkpointsInZooKeeper`, we need to adapt the client by 

```java
        // Ensure that the checkpoints path exists
		client.newNamespaceAwareEnsurePath(checkpointsPath)
				.ensure(client.getZookeeperClient());

		// All operations will have the path as root
		this.client = client.usingNamespace(client.getNamespace() + checkpointsPath);
```

calling `this(...)` should be the first statement, I don't know how to do some initialization first and then call another constructor.

Comment (TisonKun) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java]: @zentol I try to get rid of the null check by adapting the client first. However, this introduces two more `private` constructors for the unified path. I'd like to hear from your advice on this change.

**Comment by GJL** [comment]
@TisonKun Thanks for updating the PR. I will take a look after my vacation.

**Comment by TisonKun** [comment]
Sure it is up to your spare time and enjoy your vacation @GJL!

**Comment by GJL** [comment]
@TisonKun I added a commit to get rid of the ZK client in `ZooKeeeperCompletedCheckpointStore`. See #7441. I am closing this one.

**Comment by TisonKun** [comment]
@GJL thanks for your follow-up :-)
